common/transport/mqtt/src/main/java/org/thingsboard/server/transport/mqtt/MqttTransportHandler.java

```diff
@@ -375,14 +375,60 @@ private void handleGatewayPublishMsg(ChannelHandlerContext ctx, String topicName
         }
     }
 
-    private void handleSparkplugPublishMsg(ChannelHandlerContext ctx, String topicName, int msgId, MqttPublishMessage mqttMsg) {
+    private void handleSparkplugPublishMsg(ChannelHandlerContext ctx, String topicName, MqttPublishMessage mqttMsgOld) {
+        MqttPublishMessage mqttMsg = sparkplugSessionHandler.reCreateMqttPublishMessageWithPacketId(mqttMsgOld);
```

Comment:
No need to repack the message. If the QoS is 0, the msgId is -1

Reply:
Ok

Reply:
Ok